docs: rfc 0310 asset implementation (#3340)
tari-project/tari: RFC/src/RFC-0310_AssetImplementation.md

> # RFC-0310/AssetImplementation

Reviewer: Needs to be added to the summary/menu

Reviewer 2: Added

## Asset Implementation Details

![status: draft](theme/images/status-draft.svg)

> Maintainer(s): [mikethetike](https://github.com/mikethetike)

Reviewer: minor: I think the other RFCs use full name instead of handle

Reviewer 2: I don't want to add my full name if I can avoid it

Reviewer 3: Mr Tike would prefer not to be doxxed

# Licence

[ The 3-Clause BSD Licence](https://opensource.org/licenses/BSD-3-Clause).

> Copyright <YEAR> <COPYRIGHT HOLDER | The Tari Development Community>

Reviewer: placeholders

SERVICES; LOSS OF USE, DATA OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
WHETHER IN CONTRACT, STRICT LIABILITY OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

## Language

The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",
"NOT RECOMMENDED", "MAY" and "OPTIONAL" in this document are to be interpreted as described in
[BCP 14](https://tools.ietf.org/html/bcp14) (covering RFC2119 and RFC8174) when, and only when, they appear in all capitals, as
shown here.

## Disclaimer

This document and its content are intended for information purposes only and may be subject to change or update
without notice.

This document may include preliminary concepts that may or may not be in the process of being developed by the Tari
community. The release of this document is intended solely for review and discussion by the community regarding the
technological merits of the potential system outlined herein.

> ## Abstract

Reviewer: empty?

> ## Related Requests for Comment

Reviewer: ```suggestion
## Related Requests for Comment

None.
```

## Problem Statement

This RFC details the specifics of how NFTs are stored and represented on 
the base layer and the digital assets layer.

## Proposal

### Changes to outputs
The following fields must be included in the TransactionOutput, inside the OutputFeatures:

New flags to be added to OutputFlags

| Name | Value | Description |
| --- | --- | --- |

> | NON_FUNGIBLE | 0b1000_1000 | This UTXO contains Non Fungile data and must not be combined with other fungible UTXOs |

Reviewer: s/Fungile/Fungible/

Reviewer 3: These codes don't match the values in the code.

In the code, NON_FUNGIBLE is `0b1000`.

REGISTER is `0b10`, so NFT register would be `0b1010` etc

There are inconsistencies here and in the code, so we should tighten up the bitflag definitions.

| ASSET_REGISTRATION | 0b1100_1000 | This UTXO contains registration information for a new asset |

> | MINT |0b0100_0000 | This UTXO represents the creation of a new NFT, and is used during the  |

Reviewer: During the?

Reviewer 2: lol

| BURN | 0b0010_0000 | The `unique_id` in this UTXO can be spent as a normal input and must not appear in the output set unless accompanied by a `MINT` flag |
| SIDECHAIN_CHECKPOINT | 0b1001_1000 | This UTXO is a checkpoint for a sidechain | 

New fields added to `OutputFeatures`

> Note: this replaces some information in [RFC-0311](RFC-0310_AssetImplementation.md)

Reviewer: link is wrong

Reviewer 3: We also need some additional lines in the RFC metadata, like "Deprecates RFC-xxx" and "Deprecated by RFC-XXX"

> Note: this replaces some information in [RFC-0311](RFC-0310_AssetImplementation.md)

Reviewer 4: Perhaps we should add a cross-reference table to make it easy to see which fields are required for what flags.

Reviewer 3: If the bit fields are set up correctly, it should be easy to make clear definitions in the code (and this RFC).

`NFT_REGISTRATION = NON_FUNGIBLE | REGISTER`
`NFT_CHECKPOINT = NON_FUNGIBLE | CHECKPOiNT` 
etc

The bitflags crate actually has this functionality [built right in](https://docs.rs/bitflags/1.3.2/bitflags/).

| Name | Type | Description |
| --- | --- | --- |
| unique_id | bytes(32) | A unique id representing a token. This value must be present only when the NON_FUNGIBLE flag is set. The pair (parent_pub_key, unique_id) will be unique across the unspent set as described below |
| parent_pub_key | bytes(32) |  A namespacing field that also allows a heirachy of assets. In most cases this will be the asset that a token belongs to, but future uses may be possible as well |

> | asset_registration | struct | Data for registering of an asset. This value **MUST** be present if ASSET_REGISTRATION flag is set |

Reviewer 4: I am taking this as meaning that for every other flag, the data MUST not be present? Perhaps we can clarify this. A cross-reference table should make this easy.

Reviewer 2: Having data be present or not present is actually a broader scope than this specific document. By disallowing extra data from appearing, you make it difficult to soft fork in changes in future. It's probably safer to allow extra fields (protobuf serialization will ignore them). Since in another RFC it is stated that features are paid for in fees, the user can add some arbitrary data, but must 1. sign it in the metadata_signature and 2. must include it in the weighting calculation

| asset_registration.name | bytes(64) | utf8 name of the asset. Optional |
| asset_registration.template_ids | Vec<bytes(32)> |The templates that this asset exposes. Usually, at least one will be specified |
| asset_registration.creator_sig | bytes(64) (32 Nonce + 32 sig) | A signature of the UTXO's `commitment` and `script` using the public key in `unique_id` to prove this registration was created in this UTXO |
| asset_registration.checkpoint_unique_id | bytes(32) | A reference to the unique id reserved for checkpoint UTXOs. Optional |
| asset_registration.checkpoint_frequency | uint32 | The frequency, in sidechain blocks (or other measure of time the sidechain uses) in which checkpoints are created |

> | mint.issuer_proof | bytes(96) (64 nonce + 32 sig) | A _ComSig_ proof proving that the owner of the UTXO containing the asset registration created this token 

Reviewer 4: ```suggestion
| mint.issuer_proof | bytes(96) (32 nonce + 64 sig) | A _ComSig_ proof proving that the owner of the UTXO containing the asset registration created this token 
```
The nonce in a comsig is a Pedersen commitment, so only 32 bytes. But the sig is 2 scalars, hence 32 + 32 = 64.

| checkpoint.merkle_root | bytes(32) | Merkle root of the sidechain data. The format and meaning of this data is specific to the sidechain implementation |

Note: `unique_id` is not always required to be a public key

### New Base Layer Consensus rules

**New Consensus Rule**

If a UTXO is received in the mempool or block with the `MINT` bit set in the flags, the containing transaction or block is valid only if 

> no other UTXO exists in the unspent set with the same `unique_id` and `parent_public_key`. 

Reviewer 4: ```suggestion
no other UTXO exists in the unspent set with the same `unique_id`,  `parent_public_key` pair. 
```

actor is able to switch this data into a new commitment, that actor gains control of the asset or token.

Note: because this requires extra validation, the gram weight of a UTXO that carries the `ASSET_REGISTRATION` flag should be higher.

### Minting a unique token

A new token is created for an asset by creating a UTXO with the `MINT` and `NON_FUNGIBLE` flags set
and the `mint_nonfungible` feature data set.

**New Consensus Rule**

When receiving a UTXO with the `MINT` and `NON_FUNGIBLE` flags set, either the `parent_public_key` must be populated, 
or the `ASSET_REGISTRATON` flag must also be set.

Note: If the `parent_public_key` is populated and the `ASSET_REGISTRATION` flag is set, it must be treated as an asset registration 
instruction. This allows sub assets to be grouped under a parent asset.

**New Consensus Rule**

When receiving a UTXO with the `MINT` and `NON_FUNGIBLE` flags set and the `ASSET_REGISTRATION` unset, the `minting_proof_signature`
must be a valid commitment signature using the commitment of the UTXO that contains the currently valid asset registration feature

> for with the `unique_id` = `parent_public_key`. The message signed must be `Hash(commitment|script|unique_id|parent_public_key)`

Reviewer: "for with the" may be missing a word

Reviewer 5: ```suggestion
with the `unique_id`  that is the same as this minted output's `parent_public_key`. The message signed must be `Hash(commitment|script|unique_id|parent_public_key)`
```

Note: because this requires extra validation, the gram weight of a UTXO that carries the `MINT` flag should be higher.

### Transferring a unique token

The `MINT` flag is only set on the first UTXO a token appears in. When it is transferred to a new UTXO, it should only have the `NON_FUNGIBLE` flag set.

### Burning a unique token

A token can be burnt by setting the `BURN` flag and then spending the UTXO. A token with the `BURN` set should be considered as valid as long as it is in the unspent set
and is only destroyed once the UTXO is spent.

**New Consensus Rule**
When receiving a transaction or block with an input with the `BURN` flag set, there must not exist a UTXO in the output set containing the same `unique_id`,`parent_public_key` pair, unless 
that output has the `MINT` flag set.

### Sidechain Checkpoints

A sidechain checkpoint is a special type of NFT UTXO. Extra sidechain data can be specified in the `sidechain` field of the output features. Like 

> other NFT's, only one sidechain checkpoint UTXO with that unique_id can exist in the unspent set at one time. The conditions of 

Reviewer: ```suggestion
other NFTs, only one sidechain checkpoint UTXO with that unique_id can exist in the unspent set at one time. The conditions of 
```

who can spend a checkpoint are determined by the TariScript on the checkpoint. In most cases it will be an `n of m` multisig condition 
using the members of the Hotstuff committee, with `m` equal to the number of members of the committee and `n` equal to `m` minus 

> the number of failures the committee can tolerate.

Reviewer: not major - but I think `m of n` is in wider use than `n of m`

Reviewer 5: I think we have made ourselves used to saying n of m but a quick google confirms m of n is more widely used.

> The unique_id can be a specificly chosen id, but can also be a random public key. 

Reviewer: ```suggestion
The unique_id can be a specifically chosen id, but can also be a random public key. 
```

There is no limit to the number of different sidechain checkpoint tokens an asset has, but it will usually be one. 

## New Base Node GRPC methods

> A base node should expose the following methods on via GRPC:

Reviewer: ```suggestion
A base node should expose the following methods via GRPC:
```

3. Checks the caller_sig is valid
4. Attempts to execute the instruction
5. At this point the instruction may fail, based on the specific rules of the template, for example:
   1. The instruction requires the role of AssetIssuer, but the owner commitment was provided
   2. The instruction is not valid for the current state of the NFT
6. No state data must be changed in the event of a failed instruction

Instructions, failed or successful, and their results, are added to the InstructionMerkleMountainRange.
The root of the InstructionMerkleMountainRange and the root of the current state data must be hashed together and
stored with a reference to the HotstuffNode. This data will be aggregated when creating the sidechain checkpoint.

The state data must be stored as a Patricia Trie, with each key being the `unique_id` and the leaf being a serialized protobuf 
of the following schema:

```protobuf

message TokenState {
  repeated TokenStatePair pairs = 1;
}

message TokenStatePair {
  string name = 1;
  bytes value = 2;
}
```

## Checkpoints

As part of the Commit phase, a validator node includes in its vote, a signature of the current state and instruction set

> in `locked_qc` that can be used as part of the threshold signature to spend the current checkpoint. This sign the state of the chain before the current instruction proposal is executed. Before

Reviewer: ```suggestion
in `locked_qc` that can be used as part of the threshold signature to spend the current checkpoint. This signs the state of the chain before the current instruction proposal is executed. Before
```

sending out the Decide messages, the current leader looks at the base layer and determines if `checkpoint_frequency`
blocks in the sidechain have been created since the last checkpoint. If so, the leader assembles a transaction
spending the previous checkpoint using the signatures it has obtained. 

## Examples

### NFT with sidechain metadata 

Use case: I want to issue an NFT that has some immutable metadata and some mutable metadata.

> In  this example, let's pretend we are making a simple game YachtClicker in 

Reviewer: ```suggestion
In this example, let's pretend we are making a simple game called YachtClicker in 
```

which every time a user clicks on a Yacht, it increases in XP. The user can change the name of the yacht, 
but only the asset issuer can award XP based on the clicks reported in the app. 

> Template used: [EditableMetadata](linktbd) 

Reviewer: todo: linktbd

      "script": "CheckMultisig 1 of 1 <committee_pub_key1>"
   }
}
```

### Minting the yacht

In our game, we now need to create a new yacht that the player can click and start upgrading. This will be a two step process,
firstly the yacht NFT UTXO must be minted, and then it can be transferred to the new owner. 

> Note that this could be done in one step if there is an out-of-band communication that provides the 
> issuer with the commitment for the utxo

Usually this process would be done via GRPC by the software running YachtClicker, but we'll do it manually for the first token.

Again using the console wallet I run:
```
tari_console_wallet --command "mint-tokens <yacht_clicker_pub_key> yacht1"
```

This creates and broadcasts a transaction with the following UTXO:

```json
{
   TransactionOutput: {
      "features": {
         // NON_FUNGIBLE | MINT
         "flags": "0b1100_0000",
         "unique_id": "hash('yacht1')",
         "parent_public_key": "<yacht_clicker_pub_key",
         "mint": {
            "issuer_proof": "<com_sig of h(commitment |script) using asset registration commitment>"
         },
         "metadata": []
      },
      "script": "Nop"
   }
}
```

> Note that this must be caller from Asset Issuer's wallet so that it can create the commitment signature. 

Reviewer: ```suggestion
Note that this must be called from Asset Issuer's wallet so that it can create the commitment signature. 
```

shown here.

## Disclaimer

This document and its content are intended for information purposes only and may be subject to change or update
without notice.

This document may include preliminary concepts that may or may not be in the process of being developed by the Tari
community. The release of this document is intended solely for review and discussion by the community regarding the
technological merits of the potential system outlined herein.

## Abstract

## Related Requests for Comment

None.

## Problem Statement

This RFC details the specifics of how NFTs are stored and represented on 
the base layer and the digital assets layer.

## Proposal

### Changes to outputs
The following fields must be included in the TransactionOutput, inside the OutputFeatures:

New flags to be added to OutputFlags

| Name | Value | Description |
| --- | --- | --- |
| NON_FUNGIBLE | 0b1000_1000 | This UTXO contains Non Fungile data and must not be combined with other fungible UTXOs |
| ASSET_REGISTRATION | 0b1100_1000 | This UTXO contains registration information for a new asset |
| MINT |0b0100_0000 | This UTXO represents the creation of a new NFT, and is used during the  |

> | BURN | 0b0010_0000 | The `unique_id` in this UTXO can be spent as a normal input and must not appear in the output set unless accompanied by a `MINT` flag |

Reviewer 5: This description is a bit confusing, is the `mint` supposed to be `BURN`?

| SIDECHAIN_CHECKPOINT | 0b1001_1000 | This UTXO is a checkpoint for a sidechain | 

New fields added to `OutputFeatures`

Note: this replaces some information in [RFC-0311](RFC-0310_AssetImplementation.md)

| Name | Type | Description |
| --- | --- | --- |

> | unique_id | bytes(32) | A unique id representing a token. This value must be present only when the NON_FUNGIBLE flag is set. The pair (parent_pub_key, unique_id) will be unique across the unspent set as described below |

Reviewer 6: ```suggestion
| unique_id | bytes(32) | A unique id representing a token. This value MUST be present only when the NON_FUNGIBLE flag is set. The pair (parent_pub_key, unique_id) will be unique across the unspent set as described below |
```

Reviewer 3: I'd love us to look for ways to conserve precious block space. The additional fields on OutputFeatures alone run to over 384 bytes (by very quick mental math) depending on the specifics.

There are a couple of signatures and fields here that I believe, with careful consideration, could be folded into the current UTXO fields, and offer some privacy for assets to boot.

E.g. The unique id might be added to the commitment as C = kG + vH + tT for instance.

> | parent_pub_key | bytes(32) |  A namespacing field that also allows a heirachy of assets. In most cases this will be the asset that a token belongs to, but future uses may be possible as well |

Reviewer 7: ```suggestion
| parent_pub_key | bytes(32) |  A namespacing field that also allows a hierarchy of assets. In most cases this will be the asset that a token belongs to, but future uses may be possible as well |
```

| asset_registration | struct | Data for registering of an asset. This value **MUST** be present if ASSET_REGISTRATION flag is set |
| asset_registration.name | bytes(64) | utf8 name of the asset. Optional |
| asset_registration.template_ids | Vec<bytes(32)> |The templates that this asset exposes. Usually, at least one will be specified |

> | asset_registration.creator_sig | bytes(64) (32 Nonce + 32 sig) | A signature of the UTXO's `commitment` and `script` using the public key in `unique_id` to prove this registration was created in this UTXO |

Reviewer 5: does this imply that `unique_id` is a public key? It is not mentioned above.

Reviewer 5: I see it is mentioned below in more detail so perhaps just a mention of its public key nature in the table.

Reviewer 3: In the example, it's just a hash though. So not always a pubkey

| asset_registration.checkpoint_unique_id | bytes(32) | A reference to the unique id reserved for checkpoint UTXOs. Optional |
| asset_registration.checkpoint_frequency | uint32 | The frequency, in sidechain blocks (or other measure of time the sidechain uses) in which checkpoints are created |
| mint.issuer_proof | bytes(96) (64 nonce + 32 sig) | A _ComSig_ proof proving that the owner of the UTXO containing the asset registration created this token 
| checkpoint.merkle_root | bytes(32) | Merkle root of the sidechain data. The format and meaning of this data is specific to the sidechain implementation |

> Note: `unique_id` is not always required to be a public key

Reviewer 5: Elaborate? My reading is that its a public key for asset registration but any random ID for a token?

Reviewer 3: Yes, this was also unclear for me.

### New Base Layer Consensus rules

**New Consensus Rule**

If a UTXO is received in the mempool or block with the `MINT` bit set in the flags, the containing transaction or block is valid only if 
no other UTXO exists in the unspent set with the same `unique_id` and `parent_public_key`. 

**New Consensus Rule**

If a UTXO is received in the mempool or block with the `NON_FUNGIBLE` bit set and the `MINT` bit unset, the containing transaction 
or block is valid only if:
1. It contains exactly one input with the same `unique_id` and `parent_public_key`
2. It contains no other UTXOs with the same `unique_id` and `parent_public_key`

Note that the `ASSET_REGISTRATION` and `SIDE_CHAIN_CHECKPOINT` contain the `NON_FUNGIBLE` bit and will also contain 
the `MINT` bit for the first instance of each of these.

### Registering an asset

> A new asset is registered by creating a UTXO with the ASSET_REGISTRATION flag set, a new public key in `unique_id`,

Reviewer 6: Maybe expand on the usage of unique_id. If I understand correctly it can represent two types of data within the protocol. On registration, it is the public key of the registered asset and MUST be placed into the `parent_public_key` of the minted child token UTXOs. On a token UTXO it can be any sequence of bytes unique to the `parent_public_key` namespace.  Any reason we don't use a separate field for asset "class" public key?

and the data in `asset_registration` populated.

**New Consensus Rule** 

When receiving a UTXO with the `ASSET_REGISTRATION` flag set, the `creator_sig` must contain a
Schnorr signature using the public key in `unique_id` and the following challenge:

```
hash(unique_id, commitment, script)
```

This ensures that a third party cannot remove the asset registration transaction and register it 
under their UTXO. The other fields in the `OutputFeatures` struct are already locked by the `metadata_signature`
and don't need to be included.

A key part of this proposal relies on the owner of a utxo to be determined via a commitment signature, so if a malicious 
actor is able to switch this data into a new commitment, that actor gains control of the asset or token.

Note: because this requires extra validation, the gram weight of a UTXO that carries the `ASSET_REGISTRATION` flag should be higher.

### Minting a unique token

A new token is created for an asset by creating a UTXO with the `MINT` and `NON_FUNGIBLE` flags set
and the `mint_nonfungible` feature data set.

**New Consensus Rule**

> When receiving a UTXO with the `MINT` and `NON_FUNGIBLE` flags set, either the `parent_public_key` must be populated, 
> or the `ASSET_REGISTRATON` flag must also be set.

Reviewer 6: typo: `ASSET_REGISTRATON`

I found this a little clearer
```suggestion
When receiving a UTXO with the `MINT` and `NON_FUNGIBLE` flags set, the `parent_public_key` and/or the `ASSET_REGISTRATION` flag must be set.
```

Note: If the `parent_public_key` is populated and the `ASSET_REGISTRATION` flag is set, it must be treated as an asset registration 
instruction. This allows sub assets to be grouped under a parent asset.

**New Consensus Rule**

> When receiving a UTXO with the `MINT` and `NON_FUNGIBLE` flags set and the `ASSET_REGISTRATION` unset, the `minting_proof_signature`

Reviewer 5: `minting_proof_signature` is not mentioned anywhere else. Where does it live?

must be a valid commitment signature using the commitment of the UTXO that contains the currently valid asset registration feature
for with the `unique_id` = `parent_public_key`. The message signed must be `Hash(commitment|script|unique_id|parent_public_key)`

Note: because this requires extra validation, the gram weight of a UTXO that carries the `MINT` flag should be higher.

### Transferring a unique token

> The `MINT` flag is only set on the first UTXO a token appears in. When it is transferred to a new UTXO, it should only have the `NON_FUNGIBLE` flag set.

Reviewer 5: Don't we need some kind of consensus rule here to enforce that the `unique_id`s of any `NON_FUNGIBLE` flagged inputs  must appear in the outputs of this block?

Reviewer 5: I found you mention this down in the example but still think we should have a formal New Consensus Rule section here.

### Burning a unique token

A token can be burnt by setting the `BURN` flag and then spending the UTXO. A token with the `BURN` set should be considered as valid as long as it is in the unspent set
and is only destroyed once the UTXO is spent.

> **New Consensus Rule**

Reviewer 5: Do we perhaps need another consensus rule that enforces the burn that means you cannot undo the burn by spending to another output without the BURN flag?

When receiving a transaction or block with an input with the `BURN` flag set, there must not exist a UTXO in the output set containing the same `unique_id`,`parent_public_key` pair, unless 
that output has the `MINT` flag set.

### Sidechain Checkpoints

A sidechain checkpoint is a special type of NFT UTXO. Extra sidechain data can be specified in the `sidechain` field of the output features. Like 
other NFT's, only one sidechain checkpoint UTXO with that unique_id can exist in the unspent set at one time. The conditions of 
who can spend a checkpoint are determined by the TariScript on the checkpoint. In most cases it will be an `n of m` multisig condition 
using the members of the Hotstuff committee, with `m` equal to the number of members of the committee and `n` equal to `m` minus 
the number of failures the committee can tolerate.

The unique_id can be a specificly chosen id, but can also be a random public key. 

> There is no limit to the number of different sidechain checkpoint tokens an asset has, but it will usually be one. 

Reviewer 5: I agree with this, I don't think there is any value in trying to use consensus to enforce anything about the checkpoints. They are purely there as on-chain commitments that users of the sidechain can point to and cannot be used for nay kind of enforcement by the main chain nodes.

  }

```

## Validator Node Instructions

> This section is specific to the Tari Validator Node. Other sidechains may make use of 
> the base layer mechanisms for sidechains as well

Callers may issue instructions to a validator node committee member using GRPC. Instructions take the form:

```protobuf

message Instruction{
    // The public key of the asset being invoked
    bytes asset_public_key =1;
    // The template being invoked
    uint32 template_id = 2;
    // The method on the template being invoked.
    string method =3;
    // The position based arguments
    repeated bytes args = 4;
    // A commitment that exists on the base layer that identifies
    // the caller. For example, the asset issuer could use the 
    // commitment that holds the asset registration, while the owner
    // could use the commitment on the UTXO containing the NFT
    bytes caller_commitment = 4;
    // A unique number to identify this instruction
    uint64 id = 5;
    // A comsig proving the caller owns the commitment in `caller_commitment`
    // The challenge is the hash of the above fields, in order
    bytes caller_sig = 6;
    
    // Fee fields TBD
}

```

> Upon receiving an instruction, a validator node must store it in its mempool and broadcast it to all other members of

Reviewer 4: nit: Do we want to use the word mempool for the validator's memory storage of instructions?  The word mempool has become widely used in the blockchain's (aka baseLayer's) temp storage for transactions. Validator nodes technically don't deal with transactions but rather instructions. Perhals we should look into finding a unique name to help differentiate between the two?

Reviewer 2: Mempool is used in this RFC already: https://rfc.tari.com/RFC-0340_VNConsensusOverview.html#implementation

mempool and creates a proposal. 

> The leader can use any algorithm to select instructions from the mempool and it is 
valid for it to select no instructions

The proposal must conform to the below protobuf

```protobuf

message HotStuffTreeNode {
    // The parent node, as per Hotstuff Specification
    bytes parent = 1;
    
    // The payload
    InstructionSet payload = 2;
}

message InstructionSet{
   // A list of instructions
   repeated Instruction instructions = 1;
   // The hash of the block at the tip of the base layer, according to the leader's view of the base layer
   bytes base_layer_block = 3;
}

```

> Note that the base layer tip used by the leader may be a number of blocks below the actual tip to minimise the 
> number of reorgs and increase the chance that all committee members have the same view of the base layer

The committee members process the proposal via Hotstuff, except that during the prepare and pre-commit stages, they must
confirm that block in `base_layer_block` is still present in the base node that they are connected to. 

If a validator node has reached the Commit stage, it must prevent the base layer from reorging by issuing a GRPC call to 

> `LockBlock`

Reviewer 4: We need to properly think how this is implemented. This does open up a lot of attack vectors for a call like this. 

Should we not handle this the same way we handle a double spend or normal transaction for that matter. It's the validator node that cares that the checkpoint is in the main chain. A specific baseNode won't care about "Jhon's private checkpoint". If it gets reorged out they need to re-add it. Keeping old forked orphan chains breaks a lot of the block_chain ideas and needs special rules to cover those.

Reviewer 4: Lets take this example:
There are 2 NFTs, `A` and `B`. 
At block height 1 Pete has 1 utxo with funds. 
At block height 2 Pete buys a NFT `A` with those funds. 
At block height 3 we have a checkpoint confirming Pete has NFT `A`

Now the block chain reorgs back to height 1
At block height 2-1 Pete buys a NFT `B` with those funds. 
At block height 3-1 we have a checkpoint confirming Pete has NFT `B`

So now at block height 3-1 what does Pete own? NFT `A `or `B` because according to the "LockedBlock" he owns `A`, but according to the main chain he owns `B`. Or does he own both and he just claimed a free NFT? When you want to buy NFT `A`, who owns it?

According to blockchain rules, NFT `A` is the person who owned `A` at height 1. As in the history Pete never bought `A`

Reviewer 2: Let me expand a bit on the rational of it. The problem it is trying to solve is the fact that all committee members must have the same view of the blockchain when they are deciding if an instruction is valid or not. You could remove this probability by requiring commitments to be confirmed by X blocks before they can be used as a proof, but once a Hotstuff node commits to executing a set of instructions, it can't have the landscape underneath it change. It has already told the rest of the committee it will execute the instructions, so there has to be some sort of concept of locking.

Perhaps one way of doing it is to validate and pre-execute all the instructions in the Commit or PreCommit states and then the Decide phase is simply updating the state. I do think that the block that is at the tip should be communicated to the other committee nodes, so that they can confirm their view of the chain is the same

Reviewer 6: Without an in-depth/practical understanding of hotstuff, doesn't the hotstuff protocol essentially allow for this? If the nodes do not agree on the next state in the `Decide` phase, there is still a valid way to abandon the view via the `NextView` interrupt. Should the tip change for `n - f` committee members, causing the current view to become invalid, nodes could broadcast `NextView` and the committee proceeds to the next view. If this comes too late and the view is passed, the reorg would need to be detected and instructions re-proposed for the next view i.e. pulled from the unconfirmed instruction mempool in a same/similar way to base layer transactions. 

[RFC 340](https://rfc.tari.com/RFC-0340_VNConsensusOverview.html#forks-and-byzantine-failures)
> If there is a temporary network partition, the chain may fork (up to a depth of three), but the protocol guarantees safety via the voting mechanism, and the chain will reconcile once the partition resolves.

Reviewer 3: The lock_base_node command feels unnecessary. 

We could require that as part of the COMMIT step, the Hotstuff committee members include the hash of the block they're building on; and Hotstuff consensus can be reached or declined in the usual way, but I'm not really sure that even this is necessary, as others have also mentioned.

If the chain re-orgs between the time that the COMMIT step is done and a transaction is broadcast, so be it, 
it is akin to a wallet broadcasting a transaction based on a slightly outdated view of the chain. The transaction is either still fine, or it gets rejected and the wallet reacts accordingly.

lock_base_node doesn't add any value in my view, because all it does is allow the committee members to have the same (possibly incorrect) view of the chain, and by broadcast time you have the same effect, since the base node is still allowed to reorg.

If there are persistent Byzantine failures due to different members having different views of the base chain, then we could have some sort of "resync" instruction that tries to get all committee members to converge on the same view.

If a committee transaction is rejected, the correct course of action for the committee would depend on the context and would need to be spec'd out a bit more. In any event, the committee is not waiting for the tx to be mined. It will immediately move onto the next instruction batch.

but only the asset issuer can award XP based on the clicks reported in the app. 

Template used: [EditableMetadata](linktbd) 

Tools used: Wallet CLI, Collectibles CLI

For this example, let's call the asset `yacht_clicker`. 

The first step is to register the asset. 

I create a JSON file specifying the template parameters:

```json
// yacht_clicker_registration.json
{
   "name": "yacht_clicker",
   "parent_public_key": null,
   "create_initial_checkpoint": true,
   "committee": ["<committee_pub_key1>"],
   "templates": [{
     "template_id": "editable_metadata",
     "fields": [
       "issuer.num_clicks",
       "locked.uri",
       "owner.boat_name"
     ]}]
}
```

Using a wallet that has sufficient funds to pay for the registration fees, in the CLI I execute the command 

```
tari_console_wallet --command "register-asset yacht_clicker_registration.json"
```

The wallet will create a transaction containing a UTXO with the following `OutputFeatures` and publish
to a base node.

```json
{
  TransactionOutput: {
    "features": {
      "flags": "0b1100_1000", // ASSET_REGISTRATION

>       "unique_id": [

Reviewer 6: Should this be an array type?

>         generated_pubkey

Reviewer 3: Is this the same as `<yacht_clicker_pub_key>` below?

      ],
      "parent_public_key": null,
      "asset_registration": {
        "name": "yacht_clicker",
        "templates": [
          id
          for
          editable_metadata
          template
        ],
        "creator_sig": [
          signature
          of
          h(commitment
          |
          script)
        ]
      },

>       "metadata": [
>         protobuf
>         serialized
>         template
>         data
>         as
>         Vec<u8>
>       ]

Reviewer 3: ```suggestion
      "metadata": [..] // protobuf serialized template data as Vec<u8>
```

and others

Reviewer 3: `metadata` isn't mentioned anywhere prior to this. Is this the serialisation of the `fields` field above?

If so, it represents a potential scaling problem. Ignoring the fact that I'm limited by the base layer block size limit for now, what if I want to initialise a thousand fields for my asset?

    },
    "script": "Nop"
  }
}
```

This asset has a combination of data on the base layer and metadata on the second layer. Because the ownership of the token is controlled by the base layer, it is fully decentralized and censorship resistant.
The editing of metadata is done by a permissioned committee of validator nodes, which has a quicker consensus mechanism through HotstuffBFT

The `fields` specified in registration JSON file are specific to the [EditableMetadata](tbd) template and take the form 
of `<owner>.<field_name>`. We have 3 owners here, the `issuer`, `locked` and `owner`. When processing instructions, the DAN layer will check
whether the caller has access to edit these fields. More on that later.

The `create_initial_checkpoint` option instructs the wallet to create a UTXO that the committee will use to create
checkpoints

The UTXO will look as follows:

```json
{
   TransactionOutput: {
      "features": {
         // SIDECHAIN | MINT
         "flags": "0b1101_1000",
         // Special unique id,
         "unique_id": "0x0000000000000000",

>          "parent_public_key": "<yacht_clicker_pub_key",

Reviewer 6: minor:
```suggestion
         "parent_public_key": "<yacht_clicker_pub_key>",
```
Missing in a few other places
